### ISSUE TITLE
Inconsistent source end location

### ISSUE THREAD

**Issue by idoros** [postcss/postcss#1876]
There seems to be an inconsistent source mapping results for nested nodes.

The simplest example shows that the root node range starts at 0 and ends after the closing `}` while the single rule range is closing before the ending `}`

```js
const root = postcss.parse('.a{}');
/*
root.source {
    start: { offset: 0 },
    end:   { offset: 4 } // after '}'
}
root.nodes[0].source {
    start: { offset: 0 },
    end:   { offset: 3 } // before '}'
}
*/
```

With a single declaration (and no closing semicolon):
```js
const root = postcss.parse('.a{b:c}');
/*
root.source {
    start: { offset: 0 },
    end:   { offset: 7 }
}
root.nodes[0].source {
    start: { offset: 0 },
    end:   { offset: 6 } // before '}'
}
root.nodes[0].nodes[0].source {
    start: { offset: 3 },
    end:   { offset: 5 } // before 'c'
}
*/
```

With a semicolon the declaration seems to end correctly, although I would prefer to include the semicolon as part of the declaration:
```js
postcss.parse('.a{b:c;}');
/*
root.source {
    start: { offset: 0 },
    end:   { offset: 8 }
}
root.nodes[0].source {
    start: { offset: 0 },
    end:   { offset: 7 } // before '}'
}
root.nodes[0].nodes[0].source {
    start: { offset: 3 },
    end:   { offset: 6 } // before ';'
}
*/
```

With 2 declaration you can see that the second one is off and the first is "fixed" by the semicolon:
```js
postcss.parse('.a{b:c;d:e}');
/*
root.source {
    start: { offset: 0 },
    end:   { offset: 11 }
}
root.nodes[0].source {
    start: { offset: 0 },
    end:   { offset: 10 } // before '}'
}
root.nodes[0].nodes[0].source {
    start: { offset: 3 },
    end:   { offset: 6 } // before ';'
}
root.nodes[0].nodes[1].source {
    start: { offset: 7 },
    end:   { offset: 9 } // before 'e'
}
*/
```

**Comment by ai** [comment]
I am busy on another open source project and since there is no real problem, I don’t plan to spend my time.

You can investigate and send PR, I will help you.

**Comment by idoros** [comment]
I understand that a quick solution might not be readily available, and I'm perfectly fine with putting in the effort 😊.
Just wanted to make sure that I'm not missing anything and document the issue.

I will try looking into the code hopefully next week. If there are any specific tests or relevant code segments that you could direct me to, I would greatly appreciate it. Thank you in advance!"

**Comment by idoros** [comment]
I ended up looking into this yesterday after work and I think I have a solution.

Now, I could use your help on the testing front. I've thrown together some "unit" tests in a file called `location.test.js` to help me troubleshoot the different cases. 

I still have failing tests coming from [`postcss-parser-tests`](https://github.com/postcss/postcss-parser-tests) with their expected JSON need adjustments. What is the process of updating `postcss-parser-tests`? Would you like me to open a PR without addressing them for now so we can focus on the fix?

**Comment by ai** [comment]
> What is the process of updating postcss-parser-tests?

Clone `postcss` and `postcss-parser-tests` to dirs near each other:

- `Dev/`
  - `postcss/`
  - `postcss-parser-tests/`

Fix `postcss` and then:

```sh
cd ../
cd postcss-parser-tests
./test/update.js
```

Then send PR to `postcss-parser-tests` with changed JSONs.

**Comment by idoros** [comment]
done

**Comment by ai** [comment]
The fix was released at 8.4.29. Thanks.
